### ISSUE TITLE
No windows binaries for the latest release?

### ISSUE THREAD

**Issue by buckleyGI** [kmonad/kmonad#767]
Was windows support dropped recently? The latest release doesn't have the windows version.

![image](https://github.com/kmonad/kmonad/assets/2102864/b28b4143-e5f8-4265-b64f-9c5cd21ba629)

**Comment by slotThe** [comment]
I guess not :) I don't really do Windows and have no idea how one would create binaries for that platform (contributions welcome in case you do! I think one can add assets to releases after the fact). Maybe when David comes back 0.5.x can have some

**Comment by mabasic** [comment]
Can you describe the procedure to generate linux binaries so that I can try to do the same for Windows?

**Comment by slotThe** [comment]
There is a [docker container](https://github.com/kmonad/kmonad/blob/master/Dockerfile) and a [nix expression](https://github.com/kmonad/kmonad/blob/master/nix/static-compilation.nix) that are used for generating a statically linked executable (I can't remember which one I used for the new release, though :)

**Comment by mabasic** [comment]
@slotThe I have managed to build an executable for Windows... I've installed stack on windows ans used the stack build  command to generate the executable. How can we add the executable to the release?

**Comment by slotThe** [comment]
Is this executable automatically statically linked? I would assume you'd need a bit more trickery than a simple `stack build`

**Comment by mabasic** [comment]
@slotThe I have managed to build an executable for Windows... I've installed stack on windows ans used the stack build  command to generate the executable. How can we add the executable to the release?

How would I check for that? Should I move the exe to another system to verify?

**Comment by slotThe** [comment]
I don't know; on GNU/Linux one could use `ldd` to tell me whether it's a valid dynamic program, or even `file` to see whether it is recognised as being statically linked

**Comment by mabasic** [comment]
ldd output:

```
$ ldd kmonad.exe
        not a dynamic executable
```

file output:

```
$ file kmonad.exe
kmonad.exe: PE32+ executable (console) x86-64, for MS Windows
```

I've installed wsl 2 on Windows and executed those command on the previously generate exe. Is the output good?

**Comment by slotThe** [comment]
I don't know how well `file` or `ldd` know Windows binaries. Maybe we'll leave this for the next release, when there are more people around who know Windows; thanks for the effort, though :)

**Comment by clooo-jo** [comment]
> ldd output:
> 
> ```
> $ ldd kmonad.exe
>         not a dynamic executable
> ```
> 
> file output:
> 
> ```
> $ file kmonad.exe
> kmonad.exe: PE32+ executable (console) x86-64, for MS Windows
> ```
> 
> I've installed wsl 2 on Windows and executed those command on the previously generated exe. Is the output good?
> 
> I have also tried moving the file to another directory and it still works. I will try to move it to another machine also to verify.

It looks fine. Can you share the file?
I'll report whether the file works on my Windows or not.
(I'm a Windows user who needs a feature in KMonad 0.4.2)

**Comment by mabasic** [comment]
I have been using it on my two PC's for some time now. It works, but there are some things that behave differently than on Linux. Sometimes it is as CTRL gets stuck and then when you type you can do a lot of damage to the PC or the files you are working on. I can't exactly tell what triggers it. This can be reverted by pressing the win key and typing a few letters in the search bar.

Also, sometimes it seems like something else gets stuck and makes me open pinned applications from the taskbar instead of opening a new tab in the browser. This can be fixes by closing the newly opened application and continuing working.

Other than that. Sometimes when I start the application I get:

```
Encountered error in KeySink: Cannot translate to windows keycode: Missing254
Encountered error in <emitter_proc>: Cannot translate to windows keycode: Missing254
```

or

```
Encountered error in KeySource: Cannot translate from windows keycode: 6619251
```

or

```
Encountered error in KeySource: Cannot translate from windows keycode: 1634685532
```

or

```
Encountered error in KeySource: Cannot translate from windows keycode: 4522061
```

I have to kill the command because nothing works and try again until I can get the program to start without throwing that message. Then it works as usual.

As far as I can tell, the keycode changes every time I try to start the process. Just now it would not start until I have unplugged the keyboard and plugged it back again. I am on Surface Go 3 tablet, but the same thing happens on my laptop with external keyboard.

When comparing this experience with the experience on Linux, Linux is much nicer and works more stable than this, but on Windows once I get the command working it works on all keyboards. I don't have to specify the keyboard as I have to on Linux. On Linux I have a command which lets me choose the keyboard that I want and this gets tedious when I switch keyboards or take the laptop on the go. For me this is a huge benefit to using Kmonad on Windows than on Linux.

This is the executable mentioned above that I use:
https://1drv.ms/u/s!AjwIFvEX4QvYvIoYMbT7FtDm0IS24w?e=VwOlrV

**Comment by clooo-jo** [comment]
> I have been using it on my two PC's for some time now. It works, but there are some things that behave differently than on Linux. Sometimes it is as CTRL gets stuck and then when you type you can do a lot of damage to the PC or the files you are working on. I can't exactly tell what triggers it. This can be reverted by pressing the win key and typing a few letters in the search bar.
> 
> Also, sometimes it seems like something else gets stuck and makes me open pinned applications from the taskbar instead of opening a new tab in the browser. This can be fixed by closing the newly opened application and continuing working.
> 
> Other than that. Sometimes when I start the application I get:
> 
> ```
> Encountered error in KeySink: Cannot translate to windows keycode: Missing254
> Encountered error in <emitter_proc>: Cannot translate to windows keycode: Missing254
> ```
> 
> or
> 
> ```
> Encountered error in KeySource: Cannot translate from windows keycode: 6619251
> ```
> 
> or
> 
> ```
> Encountered error in KeySource: Cannot translate from windows keycode: 1634685532
> ```
> 
> or
> 
> ```
> Encountered error in KeySource: Cannot translate from windows keycode: 4522061
> ```
> 
> I have to kill the command because nothing works and try again until I can get the program to start without throwing that message. Then it works as usual.
> 
> As far as I can tell, the keycode changes every time I try to start the process. Just now it would not start until I have unplugged the keyboard and plugged it back again. I am on Surface Go 3 tablet, but the same thing happens on my laptop with external keyboard.
> 
> When comparing this experience with the experience on Linux, Linux is much nicer and works more stable than this, but on Windows once I get the command working it works on all keyboards. I don't have to specify the keyboard as I have to on Linux. On Linux I have a command which lets me choose the keyboard that I want and this gets tedious when I switch keyboards or take the laptop on the go. For me this is a huge benefit to using Kmonad on Windows than on Linux.
> 
> This is the executable mentioned above that I use: https://1drv.ms/u/s!AjwIFvEX4QvYvIoYMbT7FtDm0IS24w?e=VwOlrV

Thank you very much. It works on my PC.

Regarding the keycode error: I don't know much about how Kmonad deals with Windows keycodes but I know that some keycodes are not supported by Kmonad. I thought the problem was fixed at Kmonad 0.4.2 but it seems not to be fixed totally according to your comment.

On Kmonad 0.4.1, I can't use the keycode called "hangeul", which switches between English and Korean mode. When pressing the "hangeul" key, Kmonad prints the following message:

```
Encountered error in KeySink: Cannot translate to windows keycode: KeyHangeul
Encountered error in <emitter_proc>: Cannot translate to windows keycode: KeyHangeul
```

This error does not occur on Kmonad 0.4.2 now. (The .exe file you uploaded. Thank you again, I'm very happy with it.)

I'll use this for a few days more and report my experience.

**Comment by clooo-jo** [comment]
I had used Kmonad 0.4.2 but sometimes it works inappropriately as @mabasic addressed above.
I don't know if it is because I use Korean-typing mode (which can be turned on/off by "hangeul" key).
